searvey/usgs.py

```diff
@@ -0,0 +1,403 @@
+        raise ValueError("Product of lat range and lon range cannot exceed 25!")
+
+    usgs_stations_results = multiprocess(
+        func=_get_usgs_stations_by_output,
+        func_kwargs=[
+            {"bBox": bBox, "output": out, "hasDataType": dtp}
+            for out, dtp in product(_get_usgs_output_codes().values(), USGS_OUTPUT_TYPE)
+        ],
+    )
+    # TODO: Guard for reduce on empty total list of stations
+    usgs_stations = functools.reduce(
+        # functools.partial(pd.merge, how='outer'),
+        lambda i, j: pd.concat([i, j], ignore_index=True),
+        (r.result for r in usgs_stations_results if not r.result.empty),
+    )
+    usgs_stations = normalize_usgs_stations(usgs_stations)
+
+    usgs_stations = usgs_stations[usgs_stations.within(region)]
+
+    return usgs_stations
+
+
+def get_usgs_stations(
+    region: Optional[Union[Polygon, MultiPolygon]] = None,
+    lon_min: Optional[float] = None,
+    lon_max: Optional[float] = None,
+    lat_min: Optional[float] = None,
+    lat_max: Optional[float] = None,
+) -> gpd.GeoDataFrame:
+    """
+    Return USGS station metadata from: https://waterservices.usgs.gov/rest/Site-Service.html
+
+    If `region` is defined then the stations that are outside of the region are
+    filtered out. If the coordinates of the Bounding Box are defined then
+    stations outside of the BBox are filtered out. If both ``region`` and the
+    Bounding Box are defined, then an exception is raised.
+
+    Note: The longitudes of the USGS stations are in the [-180, 180] range.
+
+    :param region: ``Polygon`` or ``MultiPolygon`` denoting region of interest
+    :param lon_min: The minimum Longitude of the Bounding Box.
+    :param lon_max: The maximum Longitude of the Bounding Box.
+    :param lat_min: The minimum Latitude of the Bounding Box.
+    :param lat_max: The maximum Latitude of the Bounding Box.
+    :return: ``geopandas.GeoDataFrame`` with the station metadata
+    """
+    region = get_region(
+        region=region,
+        lon_min=lon_min,
+        lon_max=lon_max,
+        lat_min=lat_min,
+        lat_max=lat_max,
+        symmetric=True,
+    )
+
+    if region:
+        # NOTE: Polygon is unhashable and cannot be used for a
+        # cached function input
+        region_json = mapping(region)
+        usgs_stations = _get_usgs_stations_by_region(**region_json)
+    else:
+        usgs_stations = _get_all_usgs_stations()
+
+    return usgs_stations
+
+
+def normalize_usgs_station_data(df: pd.DataFrame, truncate_seconds: bool) -> pd.DataFrame:
+
+    # TODO: Does truncate seconds make sense for USGS?
+
+    df = df.reset_index()
+    df = df.melt(id_vars=["datetime", "site_no"], var_name="output_id")
+
+    df["qualifier"] = df.value.where(df.output_id.str.contains("_cd"))
+    df["value"] = df.value.where(~df.output_id.str.contains("_cd"))
+    df = df.dropna(subset=["value", "qualifier"], how="all")
+
+    df["code"] = df.output_id.transform(lambda i: i.split("_")[0])
```

Comment:
I haven't tested this one, so I might be missing something here, but instead of `.transform` and `lambda`, could we not use `.str` accessor methods? My understanding is that this is the ["proper" way](
https://pandas.pydata.org/docs/user_guide/text.html#string-methods) to handle string manipulation in pandas. Furthermore it would be consistent with e.g. lines 234-235 and 240-241.


For example:
``` python
df["code"] = df.output_id.str.split("_")[0]
```

Reply:
> `df["code"] = df.output_id.str.split("_")[0]`

This doesn't produce the result I'm looking for, which is: the first part of split strings from the `output_id` columns.

I think I can either use `transform` or `apply` I'm not sure how else to achieve this. The code above returns the split result of the first row instead of the first part of the split of all the rows.